File: admin/services/socket.service.js
const socketIO = require('socket.io');
const jwt = require('jsonwebtoken');
const prisma = require('../config/prisma');

let io;

/**
 * Initialiser Socket.IO
 */
const initializeSocket = (server) => {
  io = socketIO(server, {
    cors: {
      origin: process.env.ALLOWED_ORIGINS?.split(',') || '*',
      credentials: true
    }
  });

  // Middleware d'authentification
  io.use(async (socket, next) => {
    try {
      const token = socket.handshake.auth.token;
      
      if (!token) {
        return next(new Error('Authentication error'));
      }

      const decoded = jwt.verify(token, process.env.JWT_SECRET);
      
      // Récupérer l'utilisateur
      const user = await prisma.user.findUnique({
        where: { id: decoded.id },
        select: {
          id: true,
          firstName: true,
          lastName: true,
          email: true,
          avatar: true,
          role: true
        }
      });

      if (!user) {
        return next(new Error('User not found'));
      }

      socket.user = user;
      next();
    } catch (error) {
      next(new Error('Authentication error'));
    }
  });

  // Gestion des connexions
  io.on('connection', (socket) => {
    console.log(`User connected: ${socket.user.id} - ${socket.user.firstName} ${socket.user.lastName}`);

    // Rejoindre les conversations de l'utilisateur
    socket.on('join_chats', async () => {
      try {
        const chats = await prisma.chat.findMany({
          where: {
            participants: {
              has: socket.user.id
            }
          },
          select: { id: true }
        });

        chats.forEach(chat => {
          socket.join(`chat_${chat.id}`);
        });

        console.log(`User ${socket.user.id} joined ${chats.length} chat rooms`);
      } catch (error) {
        console.error('Error joining chats:', error);
      }
    });

    // Rejoindre une conversation spécifique
    socket.on('join_chat', async (chatId) => {
      try {
        const chat = await prisma.chat.findUnique({
          where: { id: parseInt(chatId) }
        });

        if (!chat) {
          socket.emit('error', { message: 'Chat not found' });
          return;
        }

        // Vérifier que l'utilisateur fait partie de la conversation
        if (!chat.participants.includes(socket.user.id)) {
          socket.emit('error', { message: 'Unauthorized' });
          return;
        }

        socket.join(`chat_${chatId}`);
        console.log(`User ${socket.user.id} joined chat ${chatId}`);
      } catch (error) {
        console.error('Error joining chat:', error);
        socket.emit('error', { message: 'Error joining chat' });
      }
    });

    // Envoyer un message
    socket.on('send_message', async (data) => {
      try {
        const { chatId, content, type = 'text' } = data;

        // Vérifier que l'utilisateur fait partie de la conversation
        const chat = await prisma.chat.findUnique({
          where: { id: parseInt(chatId) }
        });

        if (!chat) {
          socket.emit('error', { message: 'Chat not found' });
          return;
        }

        if (!chat.participants.includes(socket.user.id)) {
          socket.emit('error', { message: 'Unauthorized' });
          return;
        }

        // Créer le message
        const message = await prisma.chatMessage.create({
          data: {
            content,
            type,
            senderId: socket.user.id,
            chatId: parseInt(chatId)
          },
          include: {
            sender: {
              select: {
                id: true,
                firstName: true,
                lastName: true,
                avatar: true
              }
            }
          }
        });

        // Mettre à jour la date du chat
        await prisma.chat.update({
          where: { id: parseInt(chatId) },
          data: { updatedAt: new Date() }
        });

        // Envoyer le message à tous les participants
        io.to(`chat_${chatId}`).emit('new_message', message);

        console.log(`Message sent in chat ${chatId} by user ${socket.user.id}`);
      } catch (error) {
        console.error('Error sending message:', error);
        socket.emit('error', { message: 'Error sending message' });
      }
    });

    // Utilisateur en train d'écrire
    socket.on('typing', (chatId) => {
      socket.to(`chat_${chatId}`).emit('user_typing', {
        chatId,
        user: {
          id: socket.user.id,
          firstName: socket.user.firstName,
          lastName: socket.user.lastName
        }
      });
    });

    // Utilisateur arrête d'écrire
    socket.on('stop_typing', (chatId) => {
      socket.to(`chat_${chatId}`).emit('user_stop_typing', {
        chatId,
        userId: socket.user.id
      });
    });

    // Marquer les messages comme lus
    socket.on('mark_read', async (chatId) => {
      try {
        await prisma.chatMessage.updateMany({
          where: {
            chatId: parseInt(chatId),
            senderId: { not: socket.user.id },
            isRead: false
          },
          data: { isRead: true }
        });

        // Notifier les autres participants
        socket.to(`chat_${chatId}`).emit('messages_read', {
          chatId,
          userId: socket.user.id
        });
      } catch (error) {
        console.error('Error marking messages as read:', error);
      }
    });

    // Déconnexion
    socket.on('disconnect', () => {
      console.log(`User disconnected: ${socket.user.id}`);
    });
  });

  return io;
};

/**
 * Obtenir l'instance Socket.IO
 */
const getIO = () => {
  if (!io) {
    throw new Error('Socket.io not initialized');
  }
  return io;
};

/**
 * Envoyer une notification à un utilisateur
 */
const sendNotification = (userId, notification) => {
  if (io) {
    io.to(`user_${userId}`).emit('notification', notification);
  }
};

module.exports = {
  initializeSocket,
  getIO,
  sendNotification
};

File: yakro_actu/lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import '../models/chat_message.dart';

class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  SocketService._internal();

  IO.Socket? _socket;
  final _storage = const FlutterSecureStorage();
  
  // Callbacks
  Function(ChatMessage)? onMessageReceived;
  Function(int chatId, Map<String, dynamic> user)? onUserTyping;
  Function(int chatId, int userId)? onUserStopTyping;
  Function(int chatId, int userId)? onMessagesRead;
  Function(Map<String, dynamic>)? onNotification;
  Function(String)? onError;

  bool get isConnected => _socket?.connected ?? false;

  /// Connexion au serveur WebSocket
  Future<void> connect(String serverUrl) async {
    if (_socket != null && _socket!.connected) {
      print('Socket already connected');
      return;
    }

    try {
      // Récupérer le token
      final token = await _storage.read(key: 'access_token');
      if (token == null) {
        throw Exception('No authentication token found');
      }

      // Configuration Socket.IO
      _socket = IO.io(
        serverUrl,
        IO.OptionBuilder()
            .setTransports(['websocket'])
            .enableAutoConnect()
            .enableReconnection()
            .setAuth({'token': token})
            .build(),
      );

      // Événements de connexion
      _socket!.onConnect((_) {
        print('Socket connected');
        _joinUserChats();
      });

      _socket!.onDisconnect((_) {
        print('Socket disconnected');
      });

      _socket!.onConnectError((error) {
        print('Socket connection error: $error');
        onError?.call('Connection error: $error');
      });

      _socket!.on('error', (data) {
        print('Socket error: $data');
        onError?.call(data['message'] ?? 'Unknown error');
      });

      // Événements du chat
      _socket!.on('new_message', (data) {
        final message = ChatMessage.fromJson(data);
        onMessageReceived?.call(message);
      });

      _socket!.on('user_typing', (data) {
        onUserTyping?.call(data['chatId'], data['user']);
      });

      _socket!.on('user_stop_typing', (data) {
        onUserStopTyping?.call(data['chatId'], data['userId']);
      });

      _socket!.on('messages_read', (data) {
        onMessagesRead?.call(data['chatId'], data['userId']);
      });

      _socket!.on('notification', (data) {
        onNotification?.call(data);
      });

      // Connexion automatique
      _socket!.connect();
    } catch (e) {
      print('Error connecting socket: $e');
      rethrow;
    }
  }

  /// Rejoindre toutes les conversations de l'utilisateur
  void _joinUserChats() {
    _socket?.emit('join_chats');
  }

  /// Rejoindre une conversation spécifique
  void joinChat(int chatId) {
    if (_socket == null || !_socket!.connected) {
      print('Socket not connected');
      return;
    }
    _socket!.emit('join_chat', chatId);
  }

  /// Envoyer un message
  void sendMessage({
    required int chatId,
    required String content,
    String type = 'text',
  }) {
    if (_socket == null || !_socket!.connected) {
      print('Socket not connected');
      onError?.call('Not connected to chat server');
      return;
    }

    _socket!.emit('send_message', {
      'chatId': chatId,
      'content': content,
      'type': type,
    });
  }

  /// Notifier que l'utilisateur est en train d'écrire
  void startTyping(int chatId) {
    if (_socket == null || !_socket!.connected) return;
    _socket!.emit('typing', chatId);
  }

  /// Notifier que l'utilisateur a arrêté d'écrire
  void stopTyping(int chatId) {
    if (_socket == null || !_socket!.connected) return;
    _socket!.emit('stop_typing', chatId);
  }

  /// Marquer les messages comme lus
  void markMessagesAsRead(int chatId) {
    if (_socket == null || !_socket!.connected) return;
    _socket!.emit('mark_read', chatId);
  }

  /// Déconnexion
  void disconnect() {
    if (_socket != null) {
      _socket!.disconnect();
      _socket!.dispose();
      _socket = null;
      print('Socket disconnected and disposed');
    }
  }

  /// Reconnecter avec un nouveau token
  Future<void> reconnect(String serverUrl) async {
    disconnect();
    await connect(serverUrl);
  }
}
